runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactoryTest.java

```diff
@@ -164,7 +164,7 @@ public void createsMultipleEnvironmentOfSingleType() throws Exception {
       verify(envFactoryA, Mockito.times(0)).createEnvironment(environmentAA);
 
       bundleFactory.forStage(getExecutableStage(environmentAA));
-      verify(environmentProviderFactoryA, Mockito.times(1))
+      verify(environmentProviderFactoryA, Mockito.times(2))
```

Comment:
This does not seem correct. The factory should only be provided once.

Reply:
Thanks for your feedback! This PR allows to create multiple `ServerInfo`s and their matching `EnvironmentFactory`s when two environments have same URNs but different payloads (for example, `environment(urn: "docker", payload: "beam/python")` and `environment(urn: "docker", payload: "beam/java")`). Is there any reason I didn't know that environments with same URNs should use a single `EnvironmentFactory`?

Reply:
I think it makes sense for two Environments with the same URN also to share the same EnvironmentFactory. I don't think it is necessary to store the environment factory with the ServerInfo because there will only every be one factory anyway. We don't allow multiple EnvironmentFactories.

Reply:
I saw you changed the code to allow different environment types, e.g. Docker and Process-based environment at the same time. I don't think this is necessary because it is sufficient to have a single EnvironmentFactory (i.e. Docker) to run cross-language pipelines.

However, I don't see why we would restrict to only one environment type. So fine with me!